packages/core/src/api/app/FeatureFlags.tsx

```diff
@@ -0,0 +1,177 @@
+/*
+ * Copyright 2020 Spotify AB
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { FeatureFlagName } from '../plugin/types';
+import {
+  FeatureFlagState,
+  FeatureFlagsApi,
+} from '../apis/definitions/featureFlags';
+
+/**
+ * Helper method for validating compatibility and flag name.
+ */
+export function validateBrowserCompat(): void {
+  if (!('localStorage' in window)) {
+    throw new Error(
+      'Feature Flags are not supported on browsers without the Local Storage API',
+    );
+  }
+}
+
+export function validateFlagName(name: FeatureFlagName): void {
+  if (name.length < 3) {
+    throw new Error(`The '${name}' feature flag must have a minimum length of three characters.`);
+  }
+
+  if (name.length > 150) {
+    throw new Error(`The '${name}' feature flag must not exceed 150 characters.`);
+  }
+
+  if (!name.match(/^[a-z]+[a-z0-9-]+$/)) {
+    throw new Error(
+      `The '${name}' feature flag must start with a lowercase letter and only contain lowercase letters, numbers and hyphens. ` +
+        'Examples: feature-flag-one, alpha, release-2020',
+    );
+  }
+}
+
+/**
+ * The UserFlags class.
+ *
+ * This acts as a data structure for the user's feature flags. You
+ * can use this to retrieve, add, edit, delete, clear and save the user's
+ * feature flags to the local browser for persisted storage.
+ */
+export class UserFlags extends Map<FeatureFlagName, FeatureFlagState> {
+  constructor() {
+    validateBrowserCompat();
+
+    try {
+      const jsonString = window.localStorage.getItem('featureFlags') as string;
+      const json = JSON.parse(jsonString);
+      super(Object.entries(json));
+    } catch (err) {
+      super([]);
+    }
+  }
+
+  get(name: FeatureFlagName): FeatureFlagState {
+    validateFlagName(name);
+    return super.get(name) || FeatureFlagState.NotEnabled;
+  }
+
+  set(name: FeatureFlagName, state: FeatureFlagState): this {
+    validateFlagName(name);
+    super.set(name, state);
+    this.save();
+    return this;
+  }
+
+  delete(name: FeatureFlagName): boolean {
+    const exists = !!this.get(name);
```

Comment:
this double negation, does it play well with NotEnabled = 0? so the semantic here is not of existence, but rather whether it was set before deletion, right

Reply:
Ah this was from before where it was a `boolean` and not an enum. Fixed in 7c28c040df366965e44d151ab6d7a5d6f276e12a